### PR TITLE
Implement logging in connectors

### PR DESCRIPTION
**Pull request by LucGenetier**

### CODE REVIEW

File: src/libraries/Microsoft.PowerFx.Connectors/Environment/PowerFxConfigExtensions.cs [hunk 1]
```diff
@@ -22,24 +24,36 @@ public static class ConfigExtensions
...
-        public static IReadOnlyList<ConnectorFunction> AddActionConnector(this PowerFxConfig config, ConnectorSettings connectorSettings, OpenApiDocument openApiDocument)
+        /// <param name="configurationLogger">Logger.</param>
+        /// <returns>List of connector functions.</returns>
+        public static IReadOnlyList<ConnectorFunction> AddActionConnector(this PowerFxConfig config, ConnectorSettings connectorSettings, OpenApiDocument openApiDocument, ConnectorLogger configurationLogger = null)
         {
-            if (connectorSettings.Namespace == null)
-            {
-                throw new ArgumentNullException(nameof(connectorSettings.Namespace));
-            }
+            Guid id = ConnectorLogger.NewId();
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Environment/PowerFxConfigExtensions.cs:31]: >  Guid id = ConnectorLogger.NewId();             [](http://example.com/codeflow?start=10&length=48)

from an encapsulation perspective, consider encapsulating the id. 

var logScope = configurationLogger?.NewScope(); // this gets the id

logScope.LogInformation("message",..); // no id needed here.

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.ConfigExtensions.AddActionConnector(Microsoft.PowerFx.PowerFxConfig,Microsoft.PowerFx.Connectors.ConnectorSettings,Microsoft.OpenApi.Models.OpenApiDocument)<br />❌CP0002 M:Microsoft.PowerFx.ConfigExtensions.AddActionConnector(Microsoft.PowerFx.PowerFxConfig,System.String,Microsoft.OpenApi.Models.OpenApiDocument)<br />❌CP0002 M:Microsoft.PowerFx.Connectors.OpenApiParser.GetFunctions(Microsoft.PowerFx.Connectors.ConnectorSettings,Microsoft.OpenApi.Models.OpenApiDocument)<br />❌CP0002 M:Microsoft.PowerFx.Connectors.OpenApiParser.GetFunctions(System.String,Microsoft.OpenApi.Models.OpenApiDocument)<br />❌CP0017 M:Microsoft.PowerFx.Connectors.ConnectorFunction.GetConnectorParameterTypeAsync(Microsoft.PowerFx.Types.NamedValue[],Microsoft.PowerFx.Connectors.ConnectorParameter,Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext,System.Threading.CancellationToken)$2<br />❌CP0017 M:Microsoft.PowerFx.Connectors.ConnectorFunction.GetConnectorReturnTypeAsync(Microsoft.PowerFx.Types.NamedValue[],Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext,System.Threading.CancellationToken)$1<br />❌CP0017 M:Microsoft.PowerFx.Connectors.ConnectorFunction.GetConnectorTypeAsync(Microsoft.PowerFx.Types.NamedValue[],Microsoft.PowerFx.Connectors.ConnectorType,Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext,System.Threading.CancellationToken)$2<br />❌CP0017 M:Microsoft.PowerFx.Connectors.ConnectorFunction.InvokeAsync(Microsoft.PowerFx.Types.FormulaValue[],Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext,System.Threading.CancellationToken)$0

**Review by Robulane — APPROVED** [review]

File: src/libraries/Microsoft.PowerFx.Connectors/ConnectorHelperFunctions.cs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+            return $"{nameof(connectorSettings.Namespace)} {connectorSettings.Namespace ?? Null(nameof(connectorSettings.Namespace))}, " +
+                   $"{nameof(connectorSettings.MaxRows)} {connectorSettings.MaxRows}, " +
+                   $"{nameof(connectorSettings.IgnoreUnknownExtensions)} {connectorSettings.IgnoreUnknownExtensions}, " +
+                   $"{nameof(connectorSettings.AllowUnsupportedFunctions)} {connectorSettings.AllowUnsupportedFunctions}, ";           
+        }
+
+        internal static string LogException(Exception ex)
+        {
+            return $"Exception {ex.GetType().FullName}, Message {ex.Message}, Callstack {ex.StackTrace}";
```
[the remark this answers is not in the document]
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/ConnectorHelperFunctions.cs:140]: Have never seen any customer data in exception messages

**Comment by LucGenetier** [comment]
Public API changes: <br />❌CP0002 M:Microsoft.PowerFx.ConfigExtensions.AddActionConnector(Microsoft.PowerFx.PowerFxConfig,Microsoft.PowerFx.Connectors.ConnectorSettings,Microsoft.OpenApi.Models.OpenApiDocument)<br />❌CP0002 M:Microsoft.PowerFx.ConfigExtensions.AddActionConnector(Microsoft.PowerFx.PowerFxConfig,System.String,Microsoft.OpenApi.Models.OpenApiDocument)<br />❌CP0002 M:Microsoft.PowerFx.Connectors.OpenApiParser.GetFunctions(Microsoft.PowerFx.Connectors.ConnectorSettings,Microsoft.OpenApi.Models.OpenApiDocument)<br />❌CP0002 M:Microsoft.PowerFx.Connectors.OpenApiParser.GetFunctions(System.String,Microsoft.OpenApi.Models.OpenApiDocument)<br />❌CP0017 M:Microsoft.PowerFx.Connectors.ConnectorFunction.GetConnectorParameterTypeAsync(Microsoft.PowerFx.Types.NamedValue[],Microsoft.PowerFx.Connectors.ConnectorParameter,Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext,System.Threading.CancellationToken)$2<br />❌CP0017 M:Microsoft.PowerFx.Connectors.ConnectorFunction.GetConnectorReturnTypeAsync(Microsoft.PowerFx.Types.NamedValue[],Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext,System.Threading.CancellationToken)$1<br />❌CP0017 M:Microsoft.PowerFx.Connectors.ConnectorFunction.GetConnectorTypeAsync(Microsoft.PowerFx.Types.NamedValue[],Microsoft.PowerFx.Connectors.ConnectorType,Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext,System.Threading.CancellationToken)$2<br />❌CP0017 M:Microsoft.PowerFx.Connectors.ConnectorFunction.InvokeAsync(Microsoft.PowerFx.Types.FormulaValue[],Microsoft.PowerFx.Connectors.BaseRuntimeConnectorContext,System.Threading.CancellationToken)$0

**Pull request merged**